File: books/practical-bioinformatics/days/day-20/scripts/multi_species.py
#!/usr/bin/env python3
"""Day 20: Multi-Species Comparison — Python equivalent using requests + Biopython."""

import requests
import json
import os
import csv
from itertools import combinations
from collections import Counter

ENSEMBL_REST = "https://rest.ensembl.org"

def ensembl_symbol(species, symbol):
    """Look up gene by symbol via Ensembl REST API."""
    url = f"{ENSEMBL_REST}/lookup/symbol/{species}/{symbol}"
    resp = requests.get(url, headers={"Content-Type": "application/json"})
    resp.raise_for_status()
    return resp.json()

def ensembl_sequence(gene_id, seq_type="protein"):
    """Fetch sequence by Ensembl gene ID."""
    url = f"{ENSEMBL_REST}/sequence/id/{gene_id}"
    params = {"type": seq_type}
    resp = requests.get(url, headers={"Content-Type": "application/json"}, params=params)
    resp.raise_for_status()
    return resp.json()

def gc_content(seq):
    """Calculate GC content of a nucleotide sequence."""
    seq = seq.upper()
    gc = sum(1 for c in seq if c in "GC")
    return gc / len(seq) if len(seq) > 0 else 0.0

def kmers(seq, k):
    """Generate all k-mers from a sequence."""
    return [seq[i:i+k] for i in range(len(seq) - k + 1)]

def kmer_jaccard(seq1, seq2, k):
    """Compute Jaccard similarity of k-mer sets."""
    k1 = set(kmers(seq1, k))
    k2 = set(kmers(seq2, k))
    shared = len(k1 & k2)
    total = len(k1 | k2)
    return round(shared / total, 3) if total > 0 else 0.0

def aa_composition(seq):
    """Classify amino acids into hydrophobic, polar, charged."""
    hydrophobic = sum(1 for aa in seq if aa in "AVLIMFWP")
    polar = sum(1 for aa in seq if aa in "STNQYC")
    charged = sum(1 for aa in seq if aa in "DEKRH")
    total = len(seq)
    return {
        "hydrophobic": round(hydrophobic / total * 100, 1),
        "polar": round(polar / total * 100, 1),
        "charged": round(charged / total * 100, 1),
    }

def main():
    print("=" * 60)
    print("Day 20: Multi-Species Comparison (Python)")
    print("=" * 60)

    species = [
        {"name": "Human", "id": "homo_sapiens"},
        {"name": "Mouse", "id": "mus_musculus"},
        {"name": "Chicken", "id": "gallus_gallus"},
        {"name": "Zebrafish", "id": "danio_rerio"},
    ]

    # -- Fetch orthologs --
    print("\n-- Fetching BRCA1 Orthologs --\n")
    results = []
    for sp in species:
        try:
            gene = ensembl_symbol(sp["id"], "BRCA1")
            protein = ensembl_sequence(gene["id"], "protein")
            cds = ensembl_sequence(gene["id"], "cdna")
            prot_seq = protein["seq"]
            cds_seq = cds["seq"]
            results.append({
                "species": sp["name"],
                "gene_id": gene["id"],
                "protein_len": len(prot_seq),
                "protein_seq": prot_seq,
                "cds_len": len(cds_seq),
                "cds_seq": cds_seq,
                "gc": round(gc_content(cds_seq) * 100, 1),
            })
            print(f"  {sp['name']}: {gene['id']} ({len(prot_seq)} aa)")
        except Exception as e:
            print(f"  {sp['name']}: not found ({e})")

    # -- Comparison table --
    print("\n-- Cross-Species Comparison --\n")
    print(f"{'Species':<12} {'Protein':>8} {'CDS':>6} {'GC%':>6} {'Ratio':>6}")
    print("-" * 44)
    for r in results:
        ratio = round(r["cds_len"] / r["protein_len"], 1)
        print(f"{r['species']:<12} {r['protein_len']:>8} {r['cds_len']:>6} {r['gc']:>6} {ratio:>6}")

    # -- K-mer similarity --
    print("\n-- K-mer Jaccard Similarity (k=5) --\n")
    for i, j in combinations(range(len(results)), 2):
        sim = kmer_jaccard(results[i]["cds_seq"], results[j]["cds_seq"], 5)
        print(f"  {results[i]['species']} vs {results[j]['species']}: {sim}")

    # -- Amino acid composition --
    print("\n-- Amino Acid Composition --\n")
    for r in results:
        comp = aa_composition(r["protein_seq"])
        print(f"  {r['species']}: hydrophobic={comp['hydrophobic']}%, "
              f"polar={comp['polar']}%, charged={comp['charged']}%")

    # -- Multi-gene comparison --
    print("\n-- Multi-Gene Comparison --\n")
    genes = ["TP53", "BRCA1", "EGFR"]
    all_results = []
    for gene_symbol in genes:
        for sp in species:
            try:
                gene = ensembl_symbol(sp["id"], gene_symbol)
                prot = ensembl_sequence(gene["id"], "protein")
                all_results.append({
                    "gene": gene_symbol,
                    "species": sp["name"],
                    "length": len(prot["seq"]),
                })
            except Exception:
                pass

    print(f"{'Gene':<8} {'Species':<12} {'Length':>7}")
    print("-" * 30)
    for r in all_results:
        print(f"{r['gene']:<8} {r['species']:<12} {r['length']:>7}")

    # -- Export --
    os.makedirs("results", exist_ok=True)

    with open("results/species_comparison.csv", "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=["species", "protein_len", "cds_len", "gc_percent", "cds_protein_ratio"])
        writer.writeheader()
        for r in results:
            writer.writerow({
                "species": r["species"],
                "protein_len": r["protein_len"],
                "cds_len": r["cds_len"],
                "gc_percent": r["gc"],
                "cds_protein_ratio": round(r["cds_len"] / r["protein_len"], 1),
            })

    with open("results/brca1_orthologs.fasta", "w") as f:
        for r in results:
            f.write(f">{r['species']}_BRCA1\n{r['protein_seq']}\n")

    print("\nExported results/species_comparison.csv")
    print("Exported results/brca1_orthologs.fasta")

    print("\n" + "=" * 60)
    print("Done!")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: books/practical-bioinformatics/days/day-20/scripts/multi_species.R
#!/usr/bin/env Rscript
# Day 20: Multi-Species Comparison — R equivalent using biomaRt + httr

library(httr)
library(jsonlite)

ENSEMBL_REST <- "https://rest.ensembl.org"

ensembl_symbol <- function(species, symbol) {
  url <- paste0(ENSEMBL_REST, "/lookup/symbol/", species, "/", symbol)
  resp <- GET(url, content_type_json())
  stop_for_status(resp)
  content(resp, "parsed")
}

ensembl_sequence <- function(gene_id, type = "protein") {
  url <- paste0(ENSEMBL_REST, "/sequence/id/", gene_id, "?type=", type)
  resp <- GET(url, content_type_json())
  stop_for_status(resp)
  content(resp, "parsed")
}

gc_content <- function(seq) {
  chars <- strsplit(toupper(seq), "")[[1]]
  sum(chars %in% c("G", "C")) / length(chars)
}

kmers <- function(seq, k) {
  n <- nchar(seq)
  if (n < k) return(character(0))
  sapply(1:(n - k + 1), function(i) substr(seq, i, i + k - 1))
}

kmer_jaccard <- function(seq1, seq2, k) {
  k1 <- unique(kmers(seq1, k))
  k2 <- unique(kmers(seq2, k))
  shared <- length(intersect(k1, k2))
  total <- length(union(k1, k2))
  if (total > 0) round(shared / total, 3) else 0.0
}

aa_composition <- function(seq) {
  residues <- strsplit(seq, "")[[1]]
  total <- length(residues)
  hydrophobic <- sum(residues %in% strsplit("AVLIMFWP", "")[[1]])
  polar <- sum(residues %in% strsplit("STNQYC", "")[[1]])
  charged <- sum(residues %in% strsplit("DEKRH", "")[[1]])
  list(
    hydrophobic = round(hydrophobic / total * 100, 1),
    polar = round(polar / total * 100, 1),
    charged = round(charged / total * 100, 1)
  )
}

cat(strrep("=", 60), "\n")
cat("Day 20: Multi-Species Comparison (R)\n")
cat(strrep("=", 60), "\n")

species <- list(
  list(name = "Human", id = "homo_sapiens"),
  list(name = "Mouse", id = "mus_musculus"),
  list(name = "Chicken", id = "gallus_gallus"),
  list(name = "Zebrafish", id = "danio_rerio")
)

# -- Fetch orthologs --
cat("\n-- Fetching BRCA1 Orthologs --\n\n")
results <- list()
for (sp in species) {
  tryCatch({
    gene <- ensembl_symbol(sp$id, "BRCA1")
    protein <- ensembl_sequence(gene$id, "protein")
    cds <- ensembl_sequence(gene$id, "cdna")
    results[[length(results) + 1]] <- list(
      species = sp$name,
      gene_id = gene$id,
      protein_len = nchar(protein$seq),
      protein_seq = protein$seq,
      cds_len = nchar(cds$seq),
      cds_seq = cds$seq,
      gc = round(gc_content(cds$seq) * 100, 1)
    )
    cat(sprintf("  %s: %s (%d aa)\n", sp$name, gene$id, nchar(protein$seq)))
  }, error = function(e) {
    cat(sprintf("  %s: not found (%s)\n", sp$name, e$message))
  })
}

# -- Comparison table --
cat("\n-- Cross-Species Comparison --\n\n")
df <- data.frame(
  species = sapply(results, `[[`, "species"),
  protein_len = sapply(results, `[[`, "protein_len"),
  cds_len = sapply(results, `[[`, "cds_len"),
  gc_percent = sapply(results, `[[`, "gc"),
  cds_protein_ratio = round(sapply(results, `[[`, "cds_len") / sapply(results, `[[`, "protein_len"), 1)
)
print(df)

# -- K-mer similarity --
cat("\n-- K-mer Jaccard Similarity (k=5) --\n\n")
n <- length(results)
for (i in 1:(n - 1)) {
  for (j in (i + 1):n) {
    sim <- kmer_jaccard(results[[i]]$cds_seq, results[[j]]$cds_seq, 5)
    cat(sprintf("  %s vs %s: %s\n", results[[i]]$species, results[[j]]$species, sim))
  }
}

# -- Amino acid composition --
cat("\n-- Amino Acid Composition --\n\n")
for (r in results) {
  comp <- aa_composition(r$protein_seq)
  cat(sprintf("  %s: hydrophobic=%.1f%%, polar=%.1f%%, charged=%.1f%%\n",
              r$species, comp$hydrophobic, comp$polar, comp$charged))
}

# -- Multi-gene comparison --
cat("\n-- Multi-Gene Comparison --\n\n")
genes <- c("TP53", "BRCA1", "EGFR")
all_results <- data.frame(gene = character(), species = character(), length = integer(),
                          stringsAsFactors = FALSE)
for (gene_symbol in genes) {
  for (sp in species) {
    tryCatch({
      gene <- ensembl_symbol(sp$id, gene_symbol)
      prot <- ensembl_sequence(gene$id, "protein")
      all_results <- rbind(all_results, data.frame(
        gene = gene_symbol,
        species = sp$name,
        length = nchar(prot$seq),
        stringsAsFactors = FALSE
      ))
    }, error = function(e) {
      # Gene may not exist in all species
    })
  }
}
print(all_results)

# -- Export --
dir.create("results", showWarnings = FALSE)
write.csv(df, "results/species_comparison.csv", row.names = FALSE)

fasta_lines <- c()
for (r in results) {
  fasta_lines <- c(fasta_lines,
    paste0(">", r$species, "_BRCA1"),
    r$protein_seq
  )
}
writeLines(fasta_lines, "results/brca1_orthologs.fasta")

cat("\nExported results/species_comparison.csv\n")
cat("Exported results/brca1_orthologs.fasta\n")

cat("\n", strrep("=", 60), "\n")
cat("Done!\n")
cat(strrep("=", 60), "\n")
